vault_service/views/utils.py

```diff
@@ -85,7 +91,483 @@ def check_request(request):
+        data = data + ' database:astronomy'
+        data_list.append(data)
+
+    with current_app.session_scope() as session:
+        for d in data_list:
+            try:
+                q = session.query(MyADS).filter_by(user_id=user_id).filter_by(data=d). \
+                    filter_by(template='keyword').one()
+                current_app.logger.info('User {0} already has keyword notifications set up for keywords {1}'.
+                                        format(user_id, d))
+                existing.append({'id': q.id, 'template': 'keyword', 'name': q.name})
+            except ormexc.NoResultFound:
+                setup = MyADS(user_id=user_id,
+                              type='template',
+                              template='keyword',
+                              name=get_keyword_query_name(d),
+                              active=True,
+                              stateful=False,
+                              frequency='weekly',
+                              data=d)
+                try:
+                    session.add(setup)
+                    session.flush()
+                    myads_id = setup.id
+                    session.commit()
+                    current_app.logger.info('Added myADS keyword notifications for user {0} with keywords {1}'.
+                                            format(user_id, d))
+                except exc.IntegrityError as e:
+                    session.rollback()
+                    return json.dumps({'msg': 'New myADS setup was not saved, error: {0}'.format(e)}), 500
+
+                new.append({'id': myads_id, 'template': 'keyword', 'name': setup.name})
+
+    return existing, new
+
+
+def get_keyword_query_name(keywords):
+    """
+    For a given keyword string, return the first word or phrase, to be used in the query name
+    :param keywords: string of keywords
+    :return: first word or phrase
+    """
+    key_list = keywords.split(' ')
+    first = key_list[0]
+
+    if '"' in first and len(key_list) > 1:
+        phrase = first
+        i = 1
+        while '"' not in key_list[i]:
+            phrase = phrase + ' ' + key_list[i]
+            i += 1
+        first = phrase + ' ' + key_list[i]
+
+    if first != keywords:
+        first = first + ', etc.'
+
+    return first
+
+
+def parse_classic_keywords(query):
+    """
+    Wrapper function to parse the Classic keyword string and return a BBB-style keyword string
+
+    :param query: string; Classic-style keyword query
+    :return: string; BBB-style keyword query
+    """
+    clean_query = query.strip()
+    tree = _parse_classic_keywords_to_tree(clean_query)
+
+    v = TreeVisitor()
+    new_query = v.visit(tree).output
+
+    return new_query
+
+
+def _parse_classic_keywords_to_tree(data):
+    """
+    Given a string of keywords from Classic, parse the query tree
+
+    :param data: string of Classic keywords
+    :return: parsed tree
+    """
+
+    grammar = Lark(r"""
```

Comment:
please move `grammar` to the module scope; it will get initialized once - this operation is likely expensive